minima/database.nim

```diff
@@ -133,11 +135,43 @@ proc set*(db: Database, key: seq[byte], value: seq[byte]): Result[void, Database
 
 proc has*(db: Database, key: seq[byte]): bool =
     ## Check whether a value has been set for a key.
-    ## 
+    ##
     ## Example:
     ## 
     ## .. code-block::
     ##   let key = [byte 1, 2, 3, 4]
     ##   db.set(key, [byte 4, 3, 2, 1]])
     ##   assert(db.has(key))
     db.tree.contains(string.fromBytes(key))
+
+proc tag*(db: Database, key: seq[byte], tag: seq[byte]) =
+    ## Adds a tag to a specific key.
+    ##
+    ## Example:
+    ##
+    ## .. code-block::
+    ## let key = [byte 1, 2, 3, 4]
+    ## db.set(key, [byte 1, 2, 3, 4, 5])
+    ## db.tag(key, [byte 1, 2, 3])
+    let t = string.fromBytes(tag)
+    discard db.tags.hasKeyOrPut(t, initHashSet[seq[byte]]())
+
+    # @TODO, Tag incl must be logged
+
+    db.tags[t].incl(key)
+
+proc tags*(db: Database): seq[seq[byte]] =
+    ## Returns all the tags currently stored in the database.
+    discard
+
+iterator filter*(db: Database, tag: seq[byte]): (seq[byte], seq[byte]) =
```

Comment:
Not sure about this, it might be better to just add a streaming interface and allow piggybacking on the std library for this stuff. So for example, you can expose an `items` and `mitems` (lvalues) iterators and the standard sequtils package should be able to handle it. More info here https://nim-lang.org/docs/manual.html#iterators-and-the-for-statement and here https://nim-lang.org/docs/manual.html#iterators-and-the-for-statement-first-class-iterators.

Reply:
@dryajov yeah, I am also thinking about how to be able to do more dynamic queries. Like that I can say

```
give me all keys that are in the set x and y but not in the set z where set refers to tags.
```